File: algorithms/0019-remove-nth-node-from-end-of-list/Solution.kt
// Remove Nth Node From End - dummy + two pointers with an n+1 gap.
// Time O(n), Space O(1).

class ListNode(var `val`: Int) {
    var next: ListNode? = null
}

class Solution {
    fun removeNthFromEnd(head: ListNode?, n: Int): ListNode? {
        val dummy = ListNode(0).also { it.next = head }
        var fast: ListNode? = dummy
        var slow: ListNode? = dummy
        repeat(n + 1) {
            fast = fast?.next
        }
        while (fast != null) {
            fast = fast?.next
            slow = slow?.next
        }
        slow?.next = slow?.next?.next
        return dummy.next
    }
}

File: algorithms/0019-remove-nth-node-from-end-of-list/solution.py
# Remove Nth Node From End - dummy + two pointers with an n+1 gap.
# Time O(n), Space O(1).

class ListNode:
    def __init__(self, val: int = 0, next: "ListNode | None" = None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: ListNode | None, n: int) -> ListNode | None:
        dummy = ListNode(0, head)
        fast = slow = dummy
        for _ in range(n + 1):
            fast = fast.next
        while fast is not None:
            fast = fast.next
            slow = slow.next
        slow.next = slow.next.next
        return dummy.next
